libkbfs/disk_block_cache.go

```diff
@@ -0,0 +1,373 @@
+	}
+	return nil
+}
+
+// timeFromBytes converts a value from the LRU cache into a time.Time.
+func (*DiskBlockCacheStandard) timeFromBytes(b []byte) (t time.Time, err error) {
+	err = t.UnmarshalBinary(b)
+	return t, err
+}
+
+// getLRU retrieves the LRU time for a block in the cache, or returns
+// leveldb.ErrNotFound and a zero-valued time.Time otherwise.
+func (cache *DiskBlockCacheStandard) getLRU(tlfID tlf.ID,
+	blockID kbfsblock.ID) (time.Time, error) {
+	lruKey := cache.lruKey(tlfID, blockID.Bytes())
+	lruBytes, err := cache.lruDb.Get(lruKey, nil)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return cache.timeFromBytes(lruBytes)
+}
+
+// decodeBlockCacheEntry decodes a disk block cache entry buffer into an
+// encoded block and server half.
+func (cache *DiskBlockCacheStandard) decodeBlockCacheEntry(buf []byte) ([]byte,
+	kbfscrypto.BlockCryptKeyServerHalf, error) {
+	entry := diskBlockCacheEntry{}
+	err := cache.config.Codec().Decode(buf, &entry)
+	if err != nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
+	}
+	return entry.Buf, entry.ServerHalf, nil
+}
+
+// encodeBlockCacheEntry encodes an encoded block and serverHalf into a single
+// buffer.
+func (cache *DiskBlockCacheStandard) encodeBlockCacheEntry(buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) ([]byte, error) {
+	entry := diskBlockCacheEntry{
+		Buf:        buf,
+		ServerHalf: serverHalf,
+	}
+	return cache.config.Codec().Encode(&entry)
+}
+
+// Get implements the DiskBlockCache interface for DiskBlockCacheStandard.
+func (cache *DiskBlockCacheStandard) Get(ctx context.Context, tlfID tlf.ID,
+	blockID kbfsblock.ID) (buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf, err error) {
+	cache.lock.RLock()
+	defer cache.lock.RUnlock()
+	if cache.blockDb == nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{},
+			DiskCacheClosedError{"Get"}
+	}
+	defer func() {
+		cache.log.CDebugf(ctx, "Cache Get id=%s tlf=%s bSize=%d err=%+v",
+			blockID, tlfID, len(buf), err)
+	}()
+	blockKey := blockID.Bytes()
+	entry, err := cache.blockDb.Get(blockKey, nil)
+	if err != nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{},
+			NoSuchBlockError{blockID}
+	}
+	err = cache.updateLRULocked(ctx, tlfID, blockKey)
+	if err != nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
+	}
+	return cache.decodeBlockCacheEntry(entry)
+}
+
+// Put implements the DiskBlockCache interface for DiskBlockCacheStandard.
+func (cache *DiskBlockCacheStandard) Put(ctx context.Context, tlfID tlf.ID,
+	blockID kbfsblock.ID, buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) error {
+	cache.lock.RLock()
+	defer cache.lock.RUnlock()
+	if cache.blockDb == nil {
+		return DiskCacheClosedError{"Put"}
+	}
+	blockLen := len(buf)
+	// TODO: accounting
+	entry, err := cache.encodeBlockCacheEntry(buf, serverHalf)
+	encodeLen := len(entry)
+	cache.log.CDebugf(ctx, "Cache Put id=%s tlf=%s bSize=%d entrySize=%d err=%+v", blockID, tlfID, blockLen, encodeLen, err)
```

Comment:
is this meant to be a defer? otherwise, if any of the below run into an error, we won't know about it.